server/src/main/java/org/elasticsearch/action/bulk/BulkPrimaryExecutionContext.java

```diff
@@ -0,0 +1,321 @@
+            currentItemState != ItemProcessingState.COMPLETED :
+            currentItemState;
+        assert requestToExecute == null : requestToExecute;
+        requestToExecute = writeRequest;
+        currentItemState = ItemProcessingState.TRANSLATED;
+    }
+
+    /** returns the request that should be executed on the shard. */
+    public <T extends DocWriteRequest<T>> T getRequestToExecute() {
+        assert currentItemState == ItemProcessingState.TRANSLATED : currentItemState;
+        return (T) requestToExecute;
+    }
+
+    /** indicates that the current operation can not be completed and needs to wait for a new mapping from the master */
+    public void markAsRequiringMappingUpdate() {
+        assert currentItemState == ItemProcessingState.TRANSLATED: currentItemState;
+        currentItemState = ItemProcessingState.WAIT_FOR_MAPPING_UPDATE;
+        requestToExecute = null;
+        executionResult = null;
+    }
+
+    /** resets the current item state, prepare for a new execution */
+    public void resetForExecutionAfterRetry() {
+        assert currentItemState == ItemProcessingState.WAIT_FOR_MAPPING_UPDATE ||
+            currentItemState == ItemProcessingState.IMMEDIATE_RETRY: currentItemState;
+        currentItemState = ItemProcessingState.INITIAL;
+    }
+
+    /** completes the operation without doing anything on the primary */
+    public void markOperationAsNoOp(DocWriteResponse response) {
+        assert currentItemState != ItemProcessingState.EXECUTED &&
+            currentItemState != ItemProcessingState.COMPLETED : currentItemState;
+        assert executionResult == null;
+        executionResult = new BulkItemResponse(getCurrentItem().id(), getCurrentItem().request().opType(), response);
+        currentItemState = ItemProcessingState.EXECUTED;
+    }
+
+    /** indicates that the operation needs to be failed as the required mapping didn't arrive in time */
+    public void failOnMappingUpdate(Exception cause) {
+        assert currentItemState == ItemProcessingState.WAIT_FOR_MAPPING_UPDATE : currentItemState;
+        assert executionResult == null : executionResult;
+        currentItemState = ItemProcessingState.EXECUTED;
+        final DocWriteRequest docWriteRequest = getCurrentItem().request();
+        markAsCompleted(new BulkItemResponse(getCurrentItem().id(), docWriteRequest.opType(),
+            // Make sure to use request.index() here, if you use docWriteRequest.index() it will use the
+            // concrete index instead of an alias if used!
+            new BulkItemResponse.Failure(getCurrentItem().index(), docWriteRequest.type(), docWriteRequest.id(), cause)));
+    }
+
+    /** the current operation has been executed on the primary with the specified result */
+    public void markOperationAsExecuted(Engine.Result result) {
+        assert currentItemState == ItemProcessingState.TRANSLATED: currentItemState;
+        assert executionResult == null : executionResult;
```

Comment:
I wonder if we can add this (and similar ones) as invariant to the class (similar as was done for ReplicationTracker) we then call `assert invariant()` on each of these methods.
For example, one invariant might state that if we are in TRANSLATED state, the executionResult is null.